r/R/dplyr.R

```diff
@@ -309,8 +355,27 @@ collect.arrow_dplyr_query <- function(x, as_data_frame = TRUE, ...) {
     # See dataset.R for Dataset and Scanner(Builder) classes
     tab <- Scanner$create(x)$ToTable()
   } else {
-    # This is a Table/RecordBatch. See record-batch.R for the [ method
-    tab <- x$.data[x$filtered_rows, x$selected_columns, keep_na = FALSE]
+    # This is a Table or RecordBatch
+
+    # Filter and select the data referenced in selected columns
+    if (isTRUE(x$filtered_rows)) {
+      filter <- TRUE
+    } else {
+      filter <- eval_array_expression(x$filtered_rows, x$.data)
+    }
+    # TODO: shortcut if identical(names(x$.data), find_array_refs(x$selected_columns))?
+    tab <- x$.data[filter, find_array_refs(x$selected_columns), keep_na = FALSE]
```

Comment:
This line errors in this example, because `filter` is `NULL`:
```r
 iris %>% record_batch() %>%
   filter(if_any(ends_with("Width"), ~ . > 4))
# Error: Cannot extract rows with an object of class NULL
```

Reply:
Odd, I get

```
> iris %>% record_batch() %>%
   filter(if_any(ends_with("Width"), ~ . > 4))
Error in if_any(ends_with("Width"), ~. > 4) : 
  could not find function "if_any"
```

which is consistent with what I see from dplyr:

```
> iris %>% filter(if_any(ends_with("Width"), ~ . > 4))
Error: Problem with `filter()` input `..1`.
✖ could not find function "if_any"
ℹ Input `..1` is `if_any(ends_with("Width"), ~. > 4)`.
Run `rlang::last_error()` to see where the error occurred.
```

Reply:
What version of dplyr do you have installed? `if_any()` was added in 1.0.4.

Reply:
🙈 
```
> packageVersion("dplyr")
[1] ‘1.0.3’
```

Reply:
This also fails with 3.0.0. I made ARROW-11785 to investigate/fix that.